procrastinate/utils.py

```diff
@@ -96,15 +95,19 @@ def async_to_sync(awaitable: Callable[..., Awaitable[T]], *args, **kwargs) -> T:
     Given a callable returning an awaitable, call the callable, await it
     synchronously. Returns the result after it's done.
     """
-    return sync.async_to_sync(awaitable)(*args, **kwargs)
+
+    async def wrapper() -> T:
```

Comment:
Not sure if there is a nicer way ... just calming down the pyright gods.

Reply:
Might work if this util (as well as the async to sync) is removed. By using asyncio.to_thread directly, that should infer types correctly.

Reply:
From [the asyncio docs](https://docs.python.org/3/library/asyncio-runner.html#asyncio.run):
```
This function should be used as a main entry point for asyncio programs, and should ideally only be called once. It is recommended to use loop_factory to configure the event loop instead of policies
```

I don't think it is a good idea to hide `asyncio.run` under this util for the reason above.

Looking at where it gets used:

- https://github.com/procrastinate-org/procrastinate/blob/df98e61eadb1b6856d3ff3fda61affa61bc564b1/procrastinate/shell.py#L70

The Shell code might need to be refactored a bit to use `asyncio.run` at the top level and not within loop. I believe this is what @ewjoachim mentions in another comment.

- https://github.com/procrastinate-org/procrastinate/blob/98fd11e71c3d9a2de5f3d8fabd57722e2f2aabc1/procrastinate/connector.py#L93

If I have not missed something, there is no class inheriting the base async connector that doesn't redefine the sync operations. 
In that case, it might be those function implementations could be removed entirely. 
In the event user code was to inherit from that class for their own custom connector, it is worth adding a note in the breaking changes section for v3.

Reply:
> Might work if this util (as well as the sync_to_async) is removed. By using asyncio.run directly, that should infer types correctly.

Yes, I think so too. Those utility functions wrap only one code line each.

> The Shell code might need to be refactored a bit to use `asyncio.run` at the top level and not within loop. I believe this is what @ewjoachim mentions in another comment.

1. Yes, makes sense. I will refactor it. Another objection of @ewjoachim is (if I understood it correctly) that when you call `asyncio.run` inside a sync task, it will create a new event loop that is different from the main event loop (that of the worker). This isn't the case with asgiref, which ensures the same event loop is reused. But I think we should just add a warning to the documentation. It is how asyncio works.

> 
> * https://github.com/procrastinate-org/procrastinate/blob/98fd11e71c3d9a2de5f3d8fabd57722e2f2aabc1/procrastinate/connector.py#L93
> 
> If I have not missed something, there is no class inheriting the base async connector that doesn't redefine the sync operations. In that case, it might be those function implementations could be removed entirely. In the event user code was to inherit from that class for their own custom connector, it is worth adding a note in the breaking changes section for v3.

Ok, but let's tackle this in another PR.